mountainsort/src/mda/mdaio.cpp

```diff
@@ -247,151 +251,65 @@ long mda_read_uint16(uint16_t *data,struct MDAIO_HEADER *H,long n,FILE *input_fi
 	else return 0;
 }
 
-#define MDA_WRITE_MACRO(type1,type2) \
-type2 *tmp=(type2 *)malloc(sizeof(type2)*n); \
-for (i=0; i<n; i++) tmp[i]=(type2)data[i]; \
-long ret=fwrite(tmp,sizeof(type2),n,output_file); \
-free(tmp); \
-return ret;
-#define MDA_WRITE_MACRO_SAME(type1,type2) \
-return fwrite(data,sizeof(type1),n,output_file);
+template<typename TargetType, typename DataType>
+long mdaWriteData_impl(DataType* data, const long size, FILE* outputFile) {
+	if (is_same<DataType, TargetType>::value) {
+		return fwrite(data, sizeof(DataType), size, outputFile);
+	} else {
+		TargetType *tmp=(TargetType *)malloc(sizeof(TargetType) * size);
```

Comment:
If we're allowing C++ here, I would recommend using some more modern approach compared to a malloc-free combination.